File: curry/Views/TKNumberInputView.m
//
//  TKNumberInputView.m
//  Created by Devin Ross on 3/21/14.
//
/*
 
 curry || https://github.com/devinross/curry
 
 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:
 
 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
 
 */


#import "TKNumberInputView.h"
#import "UIDevice+TKCategory.h"
#import "TKInputKey.h"
#import "UIImage+TKCategory.h"
#import "TKFoundation.h"
#import "TKUIKit.h"

@implementation TKNumberInputView

- (instancetype) initWithFrame:(CGRect)frame{
	self = [self initWithFrame:frame withKeysModels:@[]];
	return self;
}


- (instancetype) initWithFrame:(CGRect)frame withKeysModels:(NSArray*)keys{
	frame.size = [TKInputView sizeOfKeyboardForMainScreen];
	
	self.padRect = CGRectMake(0, 0, 0.75, 1);
	
	CGFloat w = self.padRect.size.width * frame.size.width;
	CGFloat h = self.padRect.size.height * frame.size.height;
	
	CGSize s = CGSizeMake(w, h);
	self.oneKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"1" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.twoKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"2" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.threeKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"3" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.fourKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"4" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.fiveKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"5" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.sixKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"6" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.sevenKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"7" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.eightKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"8" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.nineKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"9" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	self.zeroKey =		[TKInputKey keyWithFrame:CGRectMakeWithSize( 0, 0, s) symbol:@"0" normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeDark runner:YES];
	
	NSMutableArray *ar = [NSMutableArray arrayWithArray:self.keypadKeys];
	[ar addObjectsFromArray:keys];
	[ar addObject:self.backspaceKey];
	if(!(self=[super initWithFrame:frame withKeysModels:ar])) return nil;
	return self;
}



- (void) layoutSubviews{
	[super layoutSubviews];
	
	CGRect padPer = self.padRect;
	CGSize cntSize = self.containerView.frame.size;
	CGRect padRect = CGRectMake(round(padPer.origin.x * cntSize.width), round(padPer.origin.y * cntSize.height), round(padPer.size.width * cntSize.width), round(padPer.size.height * cntSize.height));
	NSInteger w = padRect.size.width / 3;
	NSInteger h = padRect.size.height / 4;
	NSInteger xPad = 0, leftInset = 0, vertInset = 0, vertSpace = 0, bottomPad = 2;
	
	
	if([UIDevice currentDevice].padIdiom){
		xPad = 16, leftInset = 24, vertInset = 12, vertSpace = 10;
		w = (padRect.size.width - leftInset - leftInset - (xPad * 2)) / 3;
		h = (padRect.size.height - vertInset - vertInset - (vertSpace * 3)) / 4;
		bottomPad = 0;
	}
	
	CGSize s = CGSizeMake(w, h);
	self.oneKey.frame =		CGRectMakeWithSize( leftInset,				vertInset,		s);
	self.twoKey.frame =		CGRectMakeWithSize( leftInset+w+xPad,		vertInset,		s);
	self.threeKey.frame =	CGRectMakeWithSize( leftInset+w*2+xPad*2,	vertInset,		s);
	self.fourKey.frame =	CGRectMakeWithSize( leftInset,				h+vertSpace + vertInset,	s);
	self.fiveKey.frame =	CGRectMakeWithSize( leftInset+w+xPad,		h+vertSpace + vertInset,	s);
	self.sixKey.frame =		CGRectMakeWithSize( leftInset+w*2+xPad*2,	h+vertSpace + vertInset,	s);
	self.sevenKey.frame =	CGRectMakeWithSize( leftInset,				(h+vertSpace)*2 + vertInset,	s);
	self.eightKey.frame =	CGRectMakeWithSize( leftInset+w+xPad,		(h+vertSpace)*2 + vertInset,	s);
	self.nineKey.frame =	CGRectMakeWithSize( leftInset+w*2+xPad*2,	(h+vertSpace)*2 + vertInset,	s);
	self.zeroKey.frame =	CGRectMake( leftInset, h*3+vertSpace*3 + vertInset,	w*3+xPad*2, s.height);
	
	
	NSInteger minX = CGRectGetMaxX(self.threeKey.frame) + xPad;
	
	if([UIDevice currentDevice].phoneIdiom){
		self.backspaceKey.frame = CGRectMake(minX, vertInset, self.containerView.frame.size.width - minX + 1, padRect.size.height + bottomPad);
	}else{
		self.backspaceKey.frame = CGRectMake(minX, vertInset, s.width, h);
	}
	
}

- (NSArray*) keypadKeys{
	return @[self.oneKey,self.twoKey,self.threeKey,self.fourKey,self.fiveKey,self.sixKey,self.sevenKey,self.eightKey,self.nineKey,self.zeroKey];
}
- (void) touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event{
	
	if(self.selectedKey && [self.keypadKeys containsObject:self.selectedKey]){
		
		BOOL insert = YES;
		
		if(self.textField.delegate && [self.textField.delegate respondsToSelector:@selector(textField:shouldChangeCharactersInRange:replacementString:)])
			insert = [self.textField.delegate textField:self.textField shouldChangeCharactersInRange:NSMakeRange(self.textField.text.length, 0) replacementString:self.selectedKey.label.text];
		
		if(insert)
			[self.textField insertText:self.selectedKey.label.text];
		
		[super touchesCancelled:touches withEvent:event];
		return;
	}
	
	[super touchesEnded:touches withEvent:event];
	
}


@end

File: CurrySwift/Views/TKNumberInputView.swift
//
//  TKNumberInputView.swift
//  Created by Devin Ross on 3/21/14.
//
/*

curry || https://github.com/devinross/curry

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

*/

import UIKit


/** A numeric keypad built from `TKInputKey` views. */
@objc(TKNumberInputView)
open class TKNumberInputView: TKInputView {
	
	@objc open var padRect: CGRect = CGRect(x: 0, y: 0, width: 0.75, height: 1)
	
	@objc open var oneKey: TKInputKey!
	@objc open var twoKey: TKInputKey!
	@objc open var threeKey: TKInputKey!
	@objc open var fourKey: TKInputKey!
	@objc open var fiveKey: TKInputKey!
	@objc open var sixKey: TKInputKey!
	@objc open var sevenKey: TKInputKey!
	@objc open var eightKey: TKInputKey!
	@objc open var nineKey: TKInputKey!
	@objc open var zeroKey: TKInputKey!
	
	@objc open var keypadKeys: [TKInputKey] {
		[oneKey, twoKey, threeKey, fourKey, fiveKey, sixKey, sevenKey, eightKey, nineKey, zeroKey].compactMap { $0 }
	}
	
	@objc(initWithFrame:withKeysModels:)
	public override init(frame: CGRect, keysModels keys: [TKInputKey]) {
		
		var sized = frame
		sized.size = TKInputView.sizeOfKeyboardForMainScreen
		
		// The digits are built before super.init and adopted afterwards: Swift does not
		// allow touching self before the superclass initialiser runs, which the
		// Objective-C original did.
		let pad = CGRect(x: 0, y: 0, width: 0.75, height: 1)
		let size = CGSize(width: pad.width * sized.width, height: pad.height * sized.height)
		let digits = (0 ... 9).map { digit in
			TKInputKey(frame: CGRect(origin: .zero, size: size), symbol: "\(digit)",
			           normalType: .default, selectedType: .dark, runner: true)
		}
		
		// digits[0] is "0", which sits last on the keypad.
		let ordered = Array(digits[1 ... 9]) + [digits[0]]
		
		super.init(frame: sized, keysModels: ordered + keys)
		
		padRect = pad
		oneKey = digits[1];   twoKey = digits[2];   threeKey = digits[3]
		fourKey = digits[4];  fiveKey = digits[5];  sixKey = digits[6]
		sevenKey = digits[7]; eightKey = digits[8]; nineKey = digits[9]
		zeroKey = digits[0]
		
		install(key: backspaceKey)
	}
	
	public required init?(coder: NSCoder) {
		super.init(coder: coder)
	}
	
	open override func layoutSubviews() {
		super.layoutSubviews()
		
		let containerSize = containerView.frame.size
		let pad = CGRect(x: (padRect.origin.x * containerSize.width).rounded(),
		                 y: (padRect.origin.y * containerSize.height).rounded(),
		                 width: (padRect.size.width * containerSize.width).rounded(),
		                 height: (padRect.size.height * containerSize.height).rounded())
		
		var width = (pad.width / 3).rounded(.down)
		var height = (pad.height / 4).rounded(.down)
		var xPad: CGFloat = 0, leftInset: CGFloat = 0, vertInset: CGFloat = 0, vertSpace: CGFloat = 0
		var bottomPad: CGFloat = 2
		
		if UIDevice.current.padIdiom {
			xPad = 16; leftInset = 24; vertInset = 12; vertSpace = 10
			width = ((pad.width - leftInset - leftInset - (xPad * 2)) / 3).rounded(.down)
			height = ((pad.height - vertInset - vertInset - (vertSpace * 3)) / 4).rounded(.down)
			bottomPad = 0
		}
		
		let size = CGSize(width: width, height: height)
		let columns = [leftInset, leftInset + width + xPad, leftInset + width * 2 + xPad * 2]
		let rows = [vertInset, height + vertSpace + vertInset, (height + vertSpace) * 2 + vertInset]
		
		oneKey.frame   = CGRect(origin: CGPoint(x: columns[0], y: rows[0]), size: size)
		twoKey.frame   = CGRect(origin: CGPoint(x: columns[1], y: rows[0]), size: size)
		threeKey.frame = CGRect(origin: CGPoint(x: columns[2], y: rows[0]), size: size)
		fourKey.frame  = CGRect(origin: CGPoint(x: columns[0], y: rows[1]), size: size)
		fiveKey.frame  = CGRect(origin: CGPoint(x: columns[1], y: rows[1]), size: size)
		sixKey.frame   = CGRect(origin: CGPoint(x: columns[2], y: rows[1]), size: size)
		sevenKey.frame = CGRect(origin: CGPoint(x: columns[0], y: rows[2]), size: size)
		eightKey.frame = CGRect(origin: CGPoint(x: columns[1], y: rows[2]), size: size)
		nineKey.frame  = CGRect(origin: CGPoint(x: columns[2], y: rows[2]), size: size)
		
		let bottom = height * 3 + vertSpace * 3 + vertInset
		zeroKey.frame = CGRect(x: leftInset, y: bottom,
		                       width: width * 3 + xPad * 2, height: pad.height + bottomPad - bottom)
		
		let minX = threeKey.frame.maxX + xPad
		
		if UIDevice.current.userInterfaceIdiom == .phone {
			backspaceKey.frame = CGRect(x: minX, y: vertInset,
			                            width: containerView.frame.width - minX + 1,
			                            height: pad.height + bottomPad)
		} else {
			backspaceKey.frame = CGRect(x: minX, y: vertInset, width: size.width, height: height)
		}
	}
	
	open override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
		
		if let key = selectedKey, keypadKeys.contains(key) {
			insert(key: key)
			super.touchesCancelled(touches, with: event)
			return
		}
		
		super.touchesEnded(touches, with: event)
	}
	
	/// Inserts a keypad key's text, respecting the text field delegate's veto.
	func insert(key: TKInputKey) {
		
		guard let textField, let text = key.label?.text else { return }
		
		let length = (textField.text as NSString?)?.length ?? 0
		let allowed = textField.delegate?.textField?(textField,
		                                             shouldChangeCharactersIn: NSRange(location: length, length: 0),
		                                             replacementString: text) ?? true
		if allowed { textField.insertText(text) }
	}
	
}
